services/src/main/java/org/exoplatform/officeonline/WOPIService.java

```diff
@@ -380,6 +388,92 @@ public String getActionUrl(RequestInfo requestInfo, String fileId, String worksp
     }
   }
 
+  /**
+   * Rename file.
+   *
+   * @param fileId the file id
+   * @param config the config
+   * @param newTitle the new title
+   * @param lock the lock
+   * @return the string
+   * @throws BadParameterException the bad parameter exception
+   * @throws FileNotFoundException the file not found exception
+   * @throws InvalidFileNameException the invalid file name exception
+   * @throws RepositoryException the repository exception
+   * @throws LockMismatchException the lock mismatch exception
+   */
+  public String renameFile(String fileId, EditorConfig config, String newTitle, String lock) throws RepositoryException,
+                                                                                             OfficeOnlineException,
+                                                                                             LockMismatchException {
+    if (!fileId.equals(config.getFileId())) {
+      throw new BadParameterException("FileId doesn't match fileId specified in token");
+    }
+
+    newTitle = Text.escapeIllegalJcrChars(newTitle);
+    // Check and escape newTitle
+    if (StringUtils.isBlank(newTitle)) {
+      throw new InvalidFileNameException("Requested title is not allowed due to using illegal JCR chars");
+    }
+
+    NodeImpl node = (NodeImpl) nodeByUUID(config.getFileId(), config.getWorkspace());
+
+    if (!canRenameDocument(node)) {
+      throw new PermissionDeniedException("Cannot rename document. Permission denied");
+    }
+    if (node.isLocked() && !node.getLock().getLockToken().equals(lock)) {
+      throw new LockMismatchException("Given lock is different from the file lock", node.getLock().getLockToken());
+    }
+
+    Node parentNode = node.getParent();
+    if (parentNode.canAddMixin(NodetypeConstant.MIX_REFERENCEABLE)) {
+      parentNode.addMixin(NodetypeConstant.MIX_REFERENCEABLE);
+      parentNode.save();
+    }
+
+    if (!node.hasPermission(PermissionType.REMOVE)) {
+      Session systemSession = jcrService.getCurrentRepository().getSystemSession(config.getWorkspace());
```

Comment:
Why we force removal? If current user has no such rights - he should not be able to remove, it's normal behavioir. Don't need hack through system session!

Reply:
This 'hack' is used to get around the Social/ECMS bug that prohibits rename of the file that was uploaded to Activity Stream.